control-plane/api-gateway/common/translation.go

```diff
@@ -275,26 +278,65 @@ func (t ResourceTranslator) translateHTTPQueryMatch(match gwv1beta1.HTTPQueryPar
 	}
 }
 
-func (t ResourceTranslator) translateHTTPFilters(filters []gwv1beta1.HTTPRouteFilter, resourceMap *ResourceMap, namespace string) api.HTTPFilters {
-	var urlRewrite *api.URLRewrite
-	consulFilter := api.HTTPHeaderFilter{
-		Add: make(map[string]string),
-		Set: make(map[string]string),
-	}
-	var retryFilter *api.RetryFilter
-	var timeoutFilter *api.TimeoutFilter
-
+func (t ResourceTranslator) translateHTTPFilters(filters []gwv1beta1.HTTPRouteFilter, resourceMap *ResourceMap, namespace string) (api.HTTPFilters, api.HTTPResponseFilters) {
+	var (
+		urlRewrite            *api.URLRewrite
+		retryFilter           *api.RetryFilter
+		timeoutFilter         *api.TimeoutFilter
+		requestHeaderFilters  = []api.HTTPHeaderFilter{}
+		responseHeaderFilters = []api.HTTPHeaderFilter{}
+	)
+
+	// Convert Gateway API filters to portions of the Consul request and response filters.
+	// Multiple filters applying the same or conflicting operations are allowed but may
+	// result in unexpected behavior.
 	for _, filter := range filters {
 		if filter.RequestHeaderModifier != nil {
-			consulFilter.Remove = append(consulFilter.Remove, filter.RequestHeaderModifier.Remove...)
+			newFilter := api.HTTPHeaderFilter{}
 
-			for _, toAdd := range filter.RequestHeaderModifier.Add {
-				consulFilter.Add[string(toAdd.Name)] = toAdd.Value
+			if len(filter.RequestHeaderModifier.Remove) > 0 {
```

Comment:
why do we need the len checks here? won't the append effectively be a no-op if the remove slice is emtpy?

Reply:
Good point, addressing...